### PR TITLE
Zero amount invoices & debit notes

### PR DESCRIPTION
**Pull request by pnowosie**
PAY-182

I've modified `invoice_flow` example [[LINK](https://gist.github.com/pnowosie/6553be29d2cfd58c2ef5947334d719d7)],
`debit_note_flow` modified accordingly.

Result of the db query
```
 sqlite3 payment.db
SQLite version 3.22.0 2018-01-22 18:45:57
Enter ".help" for usage hints.
sqlite> SELECT id, role, status FROM pay_invoice ;
579a9f81-5fe1-4e56-abb0-e6d2674bbad0|P|SETTLED
579a9f81-5fe1-4e56-abb0-e6d2674bbad0|R|SETTLED
```

```
sqlite> SELECT id, role, status, total_amount_due FROM pay_debit_note ;
860a887f-5b0c-4dfe-a4c8-c743cbddf87c|P|SETTLED|0
860a887f-5b0c-4dfe-a4c8-c743cbddf87c|R|SETTLED|0
(...)

sqlite> SELECT * FROM pay_debit_note_event ;
860a887f-5b0c-4dfe-a4c8-c743cbddf87c|0x77bcd35db905a9c52ab17fe43dc4d91a5bdccbe0|RECEIVED|2021-02-11 14:07:45.198|
860a887f-5b0c-4dfe-a4c8-c743cbddf87c|0xbc3424be7e8e5e7f1df62ce70984a54376889b6b|ACCEPTED|2021-02-11 14:07:46.181|
860a887f-5b0c-4dfe-a4c8-c743cbddf87c|0xbc3424be7e8e5e7f1df62ce70984a54376889b6b|SETTLED|2021-02-11 14:07:46.181|
(...)
```

### CODE REVIEW

File: core/payment-driver/dummy/src/service.rs [hunk 1]
```diff
@@ -117,8 +119,9 @@ async fn schedule_payment(
     };
 
     // Spawned because calling payment service while handling a call from payment service
-    // would result in a deadlock.
+    // would result in a deadlock. We need to wait a bit, so parent scope be able to answer
     Arbiter::spawn(async move {
+        std::thread::sleep(Duration::from_millis(100));
```
Comment (maaktweluit) [inline core/payment-driver/dummy/src/service.rs:124]: small nit: would not import this at the top of the file or only 1 use
```suggestion
        std::thread::sleep(actix::clock::Duration::from_millis(100));
```
```suggestion
        std::thread::sleep(Duration::from_millis(100));
```

**Review by maaktweluit — APPROVED** [review]
LGTM!

Tested all examples and they are working, except for `debit_note_flow` on `dummy`.
Guess that was already broken so we can fix that in a separate PR.

Your gist also has a small error now:
```
curl https://gist.githubusercontent.com/pnowosie/6553be29d2cfd58c2ef5947334d719d7/raw/1276966f6cb6d7efc34b6878207a3015d353e89d/invoice_flow_example.patch | git apply -v
```

File: core/payment/src/dao/debit_note.rs [hunk 2]
```diff
@@ -236,22 +237,28 @@ impl<'c> DebitNoteDao<'c> {
...
                 .select((dsl::activity_id, dsl::total_amount_due, dsl::role))
                 .first(conn)?;
-            update_status(
-                &vec![debit_note_id.clone()],
-                &owner_id,
-                &DocumentStatus::Accepted,
-                conn,
-            )?;
+            let mut events = vec![DebitNoteEventType::DebitNoteAcceptedEvent];
+            let status = if amount.0 == BigDecimal::from(0) {
```
Comment (Wiezzel) [inline core/payment/src/dao/debit_note.rs:241]: This line could use some comment. It looks a little bit hacky and someone reading this code might get confused.

File: core/payment/src/dao/invoice.rs [hunk 3]
```diff
@@ -233,17 +233,28 @@ impl<'c> InvoiceDao<'c> {
                 .find((&invoice_id, &owner_id))
                 .select((dsl::agreement_id, dsl::amount, dsl::role))
                 .first(conn)?;
-            update_status(&invoice_id, &owner_id, &DocumentStatus::Accepted, conn)?;
+            let mut events = vec![InvoiceEventType::InvoiceAcceptedEvent];
+            let status = if amount.0 == BigDecimal::from(0) {
```
Comment (Wiezzel) [inline core/payment/src/dao/invoice.rs:237]: Ditto.

**Review by Wiezzel — APPROVED** [review]
I suggest commenting the crucial lines in the code, otherwise alright.

**Pull request merged**